ethcore/src/engines/authority_round.rs

```diff
@@ -0,0 +1,411 @@
+use std::sync::Weak;
+use std::time::{UNIX_EPOCH, Duration};
+use util::*;
+use ethkey::{verify_address, Signature};
+use rlp::{UntrustedRlp, View, encode};
+use account_provider::AccountProvider;
+use block::*;
+use spec::CommonParams;
+use engines::Engine;
+use header::Header;
+use error::{Error, BlockError};
+use evm::Schedule;
+use ethjson;
+use io::{IoContext, IoHandler, TimerToken, IoService, IoChannel};
+use service::ClientIoMessage;
+use transaction::SignedTransaction;
+use env_info::EnvInfo;
+use builtin::Builtin;
+
+/// `AuthorityRound` params.
+#[derive(Debug, PartialEq)]
+pub struct AuthorityRoundParams {
+	/// Gas limit divisor.
+	pub gas_limit_bound_divisor: U256,
+	/// Time to wait before next block or authority switching.
+	pub step_duration: Duration,
+	/// Valid authorities.
+	pub authorities: Vec<Address>,
+	/// Number of authorities.
+	pub authority_n: usize,
+}
+
+impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
+	fn from(p: ethjson::spec::AuthorityRoundParams) -> Self {
+		AuthorityRoundParams {
+			gas_limit_bound_divisor: p.gas_limit_bound_divisor.into(),
+			step_duration: Duration::from_secs(p.step_duration.into()),
+			authority_n: p.authorities.len(),
+			authorities: p.authorities.into_iter().map(Into::into).collect::<Vec<_>>(),
+		}
+	}
+}
+
+/// Engine using `AuthorityRound` proof-of-work consensus algorithm, suitable for Ethereum
+/// mainnet chains in the Olympic, Frontier and Homestead eras.
+pub struct AuthorityRound {
+	params: CommonParams,
+	our_params: AuthorityRoundParams,
+	builtins: BTreeMap<Address, Builtin>,
+	transistion_service: IoService<BlockArrived>,
+	message_channel: Mutex<Option<IoChannel<ClientIoMessage>>>,
+	step: AtomicUsize,
+	proposed: AtomicBool,
+}
+
+fn header_step(header: &Header) -> Result<usize, ::rlp::DecoderError> {
+	UntrustedRlp::new(&header.seal()[0]).as_val()
+}
+
+fn header_signature(header: &Header) -> Result<Signature, ::rlp::DecoderError> {
+  UntrustedRlp::new(&header.seal()[1]).as_val::<H520>().map(Into::into)
+}
+
+trait AsMillis {
+	fn as_millis(&self) -> u64;
+}
+
+impl AsMillis for Duration {
+	fn as_millis(&self) -> u64 {
+		self.as_secs()*1_000 + (self.subsec_nanos()/1_000_000) as u64
+	}
+}
+
+impl AuthorityRound {
+	/// Create a new instance of AuthorityRound engine.
+	pub fn new(params: CommonParams, our_params: AuthorityRoundParams, builtins: BTreeMap<Address, Builtin>) -> Result<Arc<Self>, Error> {
+		let initial_step = (unix_now().as_secs() / our_params.step_duration.as_secs()) as usize;
+		let engine = Arc::new(
+			AuthorityRound {
+				params: params,
+				our_params: our_params,
+				builtins: builtins,
+				transistion_service: IoService::<BlockArrived>::start()?,
+				message_channel: Mutex::new(None),
+				step: AtomicUsize::new(initial_step),
+				proposed: AtomicBool::new(false)
+			});
+		let handler = TransitionHandler { engine: Arc::downgrade(&engine) };
+		engine.transistion_service.register_handler(Arc::new(handler))?;
```

Comment:
spoke to @arkpar about this and we decided to hold off until the next release to start using `?` to maintain a bit of backwards compatibility.